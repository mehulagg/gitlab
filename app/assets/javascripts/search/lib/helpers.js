import Vue from 'vue';

export const mountComponent = (vueInstance, component, propsData, id) => {
  const ComponentClass = Vue.extend(component);
  const instance = new ComponentClass({ propsData, store: vueInstance.$store });
  instance.$mount(vueInstance.$el.querySelector(id));

  return instance;
};
