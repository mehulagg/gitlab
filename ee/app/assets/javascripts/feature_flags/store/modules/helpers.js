import _ from 'underscore';
import {
  ROLLOUT_STRATEGY_ALL_USERS,
  ROLLOUT_STRATEGY_PERCENT_ROLLOUT,
  INTERNAL_ID_PREFIX,
  DEFAULT_PERCENT_ROLLOUT,
  PERCENT_ROLLOUT_GROUP_ID,
} from '../../constants';

/**
 * Converts raw scope objects fetched from the API into an array of scope
 * objects that is easier/nicer to bind to in Vue.
 * @param {Array} scopesFromRails An array of scope objects fetched from the API
 */
export const mapToScopesViewModel = scopesFromRails =>
  (scopesFromRails || []).map(s => {
    const [strategy] = s.strategies || [];

    const rolloutStrategy = strategy ? strategy.name : ROLLOUT_STRATEGY_ALL_USERS;

    let rolloutPercentage = DEFAULT_PERCENT_ROLLOUT;
    if (strategy && strategy.parameters && strategy.parameters.percentage) {
      rolloutPercentage = strategy.parameters.percentage;
    }

    return {
      id: s.id,
      environmentScope: s.environment_scope,
      active: Boolean(s.active),
      canUpdate: Boolean(s.can_update),
      protected: Boolean(s.protected),
      rolloutStrategy,
      rolloutPercentage,

      // eslint-disable-next-line no-underscore-dangle
      shouldBeDestroyed: Boolean(s._destroy),
    };
  });

/**
 * Converts the parameters emitted by the Vue component into
 * the shape that the Rails API expects.
 * @param {Array} scopesFromVue An array of scope objects from the Vue component
 */
export const mapFromScopesViewModel = params => {
  const scopes = (params.scopes || []).map(s => {
    const parameters = {};
    if (s.rolloutStrategy === ROLLOUT_STRATEGY_PERCENT_ROLLOUT) {
      parameters.groupId = PERCENT_ROLLOUT_GROUP_ID;
      parameters.percentage = s.rolloutPercentage;
    }

    // Strip out any internal IDs
    const id = _.isString(s.id) && s.id.startsWith(INTERNAL_ID_PREFIX) ? undefined : s.id;

    return {
      id,
      environment_scope: s.environmentScope,
      active: s.active,
      can_update: s.canUpdate,
      protected: s.protected,
      _destroy: s.shouldBeDestroyed,
      strategies: [
        {
          name: s.rolloutStrategy,
          parameters,
        },
      ],
    };
  });

  return {
    operations_feature_flag: {
      name: params.name,
      description: params.description,
      scopes_attributes: scopes,
    },
  };
};

/**
 * Creates a new feature flag environment scope object for use
 * in a Vue component.  An optional parameter can be passed to
 * override the property values that are created by default.
 *
 * @param {Object} overrides An optional object whose
 * property values will be used to override the default values.
 *
 */
export const createNewEnvironmentScope = (overrides = {}) => {
  const defaultScope = {
    environmentScope: '',
    active: false,
    id: _.uniqueId(INTERNAL_ID_PREFIX),
    rolloutStrategy: ROLLOUT_STRATEGY_ALL_USERS,
    rolloutPercentage: DEFAULT_PERCENT_ROLLOUT,
  };

  const newScope = {
    ...defaultScope,
    ...overrides,
  };

  if (gon && gon.features && gon.features.featureFlagPermissions) {
    newScope.canUpdate = true;
    newScope.protected = false;
  }

  return newScope;
};
