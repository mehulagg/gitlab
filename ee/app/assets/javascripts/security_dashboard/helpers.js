import isPlainObject from 'lodash/isPlainObject';
import { ALL, BASE_FILTERS } from 'ee/security_dashboard/store/modules/filters/constants';
import { REPORT_TYPES, SEVERITY_LEVELS } from 'ee/security_dashboard/store/constants';
import { VULNERABILITY_STATES } from 'ee/vulnerabilities/constants';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { s__, __ } from '~/locale';

const parseOptions = obj =>
  Object.entries(obj).map(([id, name]) => ({ id: id.toUpperCase(), name }));

export const mapProjects = projects =>
  projects.map(p => ({ id: p.id.split('/').pop(), name: p.name }));

export const stateFilter = {
  name: s__('SecurityReports|Status'),
  id: 'state',
  allOption: { id: ALL, name: s__('VulnerabilityStatusTypes|All') },
  options: parseOptions(VULNERABILITY_STATES),
};

export const severityFilter = {
  name: s__('SecurityReports|Severity'),
  id: 'severity',
  allOption: BASE_FILTERS.severity,
  options: parseOptions(SEVERITY_LEVELS),
};

const getScannerFilterOptions = () => {
  const options = parseOptions(REPORT_TYPES);
  return options.map(option => ({
    ...option,
    queryId: `GitLab:${option.id}`,
    externalId: 'NO_SCANNERS_FOUND',
  }));
};

export const scannerFilter = {
  name: s__('Reports|Scanner'),
  id: 'reportType',
  allOption: BASE_FILTERS.report_type,
  options: getScannerFilterOptions(),
};

export const getProjectFilter = projects => {
  if (!projects?.length) return undefined;

  return {
    name: s__('SecurityReports|Project'),
    id: 'projectId',
    allOption: BASE_FILTERS.project_id,
    options: mapProjects(projects),
  };
};

/**
 * Provided a security reports summary from the GraphQL API, this returns an array of arrays
 * representing a properly formatted report ready to be displayed in the UI. Each sub-array consists
 * of the user-friend report's name, and the summary's payload. Note that summary entries are
 * considered empty and are filtered out of the return if the payload is `null` or don't include
 * a vulnerabilitiesCount property. Report types whose name can't be matched to a user-friendly
 * name are filtered out as well.
 *
 * Take the following summary for example:
 * {
 *   containerScanning: { vulnerabilitiesCount: 123 },
 *   invalidReportType: { vulnerabilitiesCount: 123 },
 *   dast: null,
 * }
 *
 * The formatted summary would look like this:
 * [
 *   ['containerScanning', { vulnerabilitiesCount: 123 }]
 * ]
 *
 * Note that `invalidReportType` was filtered out as it can't be matched with a user-friendly name,
 * and the DAST report was omitted because it's empty (`null`).
 *
 * @param {Object} rawSummary
 * @returns {Array}
 */
export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter(entry => entry !== null);
};

/**
 * We have disabled loading hasNextPage from GraphQL as it causes timeouts in database,
 * instead we have to calculate that value based on the existence of endCursor. When endCursor
 * is empty or has null value, that means that there is no next page to be loaded from GraphQL API.
 *
 * @param {Object} pageInfo
 * @returns {Object}
 */
export const preparePageInfo = pageInfo => {
  return { ...pageInfo, hasNextPage: Boolean(pageInfo?.endCursor) };
};

export const createProjectLoadingError = () => __('An error occurred while retrieving projects.');

export default () => ({});
