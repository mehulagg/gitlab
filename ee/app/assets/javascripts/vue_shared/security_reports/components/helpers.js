import { EMPTY_BODY_MESSAGE } from './constants';

export const bodyWithFallBack = (body) => {
  // Return string if defined or blank string, otherwise fallback
  return body === '' ? '' : body || EMPTY_BODY_MESSAGE;
};
