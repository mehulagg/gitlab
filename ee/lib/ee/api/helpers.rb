module EE
  module API
    module Helpers
      extend ::Gitlab::Utils::Override

      override :current_user
      def current_user
        strong_memoize(:current_user) do
          user = super

          if user
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :user, user.id)
          end

          user
        end
      end

      def check_project_feature_available!(feature)
        not_found! unless user_project.feature_available?(feature)
      end
    end
  end
end
