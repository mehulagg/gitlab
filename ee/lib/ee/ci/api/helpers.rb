module EE
  module Ci
    module API
      module Helpers
        def authenticate_build!
          id = params[:id]

          if id
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :build, id)
          end

          super
        end

        def current_runner
          token = params[:token]

          if token
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :runner, token)
          end

          super
        end
      end
    end
  end
end
