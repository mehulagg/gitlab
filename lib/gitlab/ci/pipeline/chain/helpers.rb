# frozen_string_literal: true

module Gitlab
  module Ci
    module Pipeline
      module Chain
        module Helpers
          def skip
            @pipeline.skip if persistent_rule.can_persist_skip?
          end

          ##
          # failure_reason must be one of Ci::Pipeline.failure_reasons.
          def error(message, failure_reason = nil)
            if failure_reason && persistent_rule.can_persist_error?(failure_reason)
              pipeline.yaml_errors = message if failure_reason == :config_error
              pipeline.drop!(failure_reason)
            end

            pipeline.errors.add(:base, message)
          end

          private

          def persistent_rule
            @persistent_rule ||= Chain::PersistentRule.fabricate(@pipeline.source)
          end
        end
      end
    end
  end
end
