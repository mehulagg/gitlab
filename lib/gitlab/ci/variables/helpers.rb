# frozen_string_literal: true

module Gitlab
  module Ci
    module Variables
      module Helpers
        class << self
          def merge_variables(current_vars, new_vars)
            current_vars = transform_from_yaml_variables(current_vars)
            new_vars = transform_from_yaml_variables(new_vars)

            transform_to_yaml_variables(
              current_vars.merge(new_vars)
            )
          end

          def transform_to_yaml_variables(vars)
            vars.to_h.map do |key, value|
              { key: key.to_s, value: value, public: true }
            end
          end

          def transform_from_yaml_variables(vars)
            return vars.stringify_keys if vars.is_a?(Hash)

            vars.to_a.to_h { |var| [var[:key].to_s, var[:value]] }
          end
        end
      end
    end
  end
end
