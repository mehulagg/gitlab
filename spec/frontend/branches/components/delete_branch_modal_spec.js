import { GlButton, GlModal, GlFormInput } from '@gitlab/ui';
import { shallowMount } from '@vue/test-utils';
import { stubComponent } from 'helpers/stub_component';
import DeleteBranchModal from '~/branches/components/delete_branch_modal.vue';

let wrapper;

const branchName = 'test_modal';

const createComponent = (data = {}) => {
  wrapper = shallowMount(DeleteBranchModal, {
    data() {
      return {
        branchName,
        deletePath: '/path/to/branch',
        defaultBranchName: 'default',
        ...data,
      };
    },
    attrs: {
      visible: true,
    },
    stubs: {
      GlModal: stubComponent(GlModal, {
        template:
          '<div><slot name="modal-title"></slot><slot></slot><slot name="modal-footer"></slot></div>',
      }),
      GlButton,
      GlFormInput,
    },
  });
};

const findDeleteButton = () => wrapper.find('[data-testid="delete_branch_confirmation_button"]');
const findFormInput = () => wrapper.findComponent(GlFormInput);

describe('Delete branch modal', () => {
  afterEach(() => {
    wrapper.destroy();
  });

  describe('Deleting a regular branch', () => {
    beforeEach(() => {
      createComponent();
    });

    it('renders the modal correctly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('submits the form when clicked', () => {
      const submitFormSpy = jest.spyOn(wrapper.vm.$refs.form, 'submit');

      return wrapper.vm.$nextTick().then(() => {
        findDeleteButton().trigger('click');

        expect(submitFormSpy).toHaveBeenCalled();
      });
    });
  });

  describe('Deleting a protected branch (for owner or maintainer)', () => {
    beforeEach(() => {
      createComponent({ isProtectedBranch: true });
    });

    it('renders the modal correctly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('disables the delete button when branch name input is unconfirmed', () => {
      expect(findDeleteButton().attributes('disabled')).toBe('true');
    });

    it('enables the delete button when branch name input is confirmed', () => {
      return wrapper.vm
        .$nextTick()
        .then(() => {
          findFormInput().vm.$emit('input', branchName);
        })
        .then(() => {
          expect(findDeleteButton()).not.toBeDisabled();
        });
    });
  });
});
