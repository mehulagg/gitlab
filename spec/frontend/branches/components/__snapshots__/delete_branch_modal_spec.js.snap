// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete branch modal Deleting a protected branch (for owner or maintainer) renders the modal correctly 1`] = `
"<div visible=\\"visible\\">
  <gl-alert-stub title=\\"\\" dismisslabel=\\"Dismiss\\" variant=\\"danger\\" primarybuttonlink=\\"\\" primarybuttontext=\\"\\" secondarybuttonlink=\\"\\" secondarybuttontext=\\"\\" class=\\"gl-mb-5\\">
    <div data-testid=\\"modal-message\\">
      <gl-sprintf-stub message=\\"You're about to permanently delete the protected branch %{strongStart}test_modal.%{strongEnd}\\"></gl-sprintf-stub>
      <p class=\\"gl-mb-0 gl-mt-4\\">
        This branch hasn’t been merged into default. To avoid data loss, consider merging this branch before deleting it.
      </p>
    </div>
  </gl-alert-stub>
  <form action=\\"/path/to/branch\\" method=\\"post\\">
    <div class=\\"gl-mt-4\\">
      <p>
        <gl-sprintf-stub message=\\"Once you confirm and press %{strongStart}Yes, delete protected branch,%{strongEnd} it cannot be undone or recovered.\\"></gl-sprintf-stub>
      </p>
      <p>
        <gl-sprintf-stub message=\\"Please type the following to confirm:\\"></gl-sprintf-stub> <code class=\\"gl-white-space-pre-wrap\\"> test_modal </code>
        <b-form-input-stub name=\\"delete_branch_input\\" value=\\"\\" autocomplete=\\"off\\" debounce=\\"0\\" type=\\"text\\" aria-labelledby=\\"input-label\\" class=\\"gl-form-input gl-mt-4\\"></b-form-input-stub>
      </p>
    </div> <input type=\\"hidden\\" name=\\"_method\\" value=\\"delete\\"> <input type=\\"hidden\\" name=\\"authenticity_token\\">
  </form>
  <div class=\\"gl-display-flex gl-flex-direction-row gl-justify-content-end gl-flex-wrap gl-m-0\\">
    <b-button-stub size=\\"md\\" variant=\\"default\\" type=\\"button\\" tag=\\"button\\" class=\\"gl-button\\">
      <!---->
      <!----> <span class=\\"gl-button-text\\">
        Cancel, keep branch
      </span></b-button-stub>
    <div class=\\"gl-mr-3\\"></div>
    <b-button-stub disabled=\\"true\\" size=\\"md\\" variant=\\"danger\\" type=\\"button\\" tag=\\"button\\" data-qa-selector=\\"delete_branch_confirmation_button\\" data-testid=\\"delete_branch_confirmation_button\\" class=\\"gl-button\\">
      <!---->
      <!----> <span class=\\"gl-button-text\\">Yes, delete protected branch</span></b-button-stub>
  </div>
</div>"
`;

exports[`Delete branch modal Deleting a regular branch renders the modal correctly 1`] = `
"<div visible=\\"visible\\">
  <gl-alert-stub title=\\"\\" dismisslabel=\\"Dismiss\\" variant=\\"danger\\" primarybuttonlink=\\"\\" primarybuttontext=\\"\\" secondarybuttonlink=\\"\\" secondarybuttontext=\\"\\" class=\\"gl-mb-5\\">
    <div data-testid=\\"modal-message\\">
      <gl-sprintf-stub message=\\"You're about to permanently delete the branch %{strongStart}test_modal.%{strongEnd}\\"></gl-sprintf-stub>
      <p class=\\"gl-mb-0 gl-mt-4\\">
        This branch hasn’t been merged into default. To avoid data loss, consider merging this branch before deleting it.
      </p>
    </div>
  </gl-alert-stub>
  <form action=\\"/path/to/branch\\" method=\\"post\\">
    <div>
      <p class=\\"gl-mt-4\\">
        <gl-sprintf-stub message=\\"Deleting the %{strongStart}test_modal%{strongEnd} branch cannot be undone. Are you sure?\\"></gl-sprintf-stub>
      </p>
    </div> <input type=\\"hidden\\" name=\\"_method\\" value=\\"delete\\"> <input type=\\"hidden\\" name=\\"authenticity_token\\">
  </form>
  <div class=\\"gl-display-flex gl-flex-direction-row gl-justify-content-end gl-flex-wrap gl-m-0\\">
    <b-button-stub size=\\"md\\" variant=\\"default\\" type=\\"button\\" tag=\\"button\\" class=\\"gl-button\\">
      <!---->
      <!----> <span class=\\"gl-button-text\\">
        Cancel, keep branch
      </span></b-button-stub>
    <div class=\\"gl-mr-3\\"></div>
    <b-button-stub size=\\"md\\" variant=\\"danger\\" type=\\"button\\" tag=\\"button\\" data-qa-selector=\\"delete_branch_confirmation_button\\" data-testid=\\"delete_branch_confirmation_button\\" class=\\"gl-button\\">
      <!---->
      <!----> <span class=\\"gl-button-text\\">Yes, delete branch</span></b-button-stub>
  </div>
</div>"
`;
